middleware/http_tracing.go

```diff
@@ -50,3 +46,111 @@ func (t Tracer) Wrap(next http.Handler) http.Handler {
+//
+// The httpgrpc client wraps HTTP requests up into a generic httpgrpc.HTTP/Handle gRPC method.
+// The httpgrpc server unwraps httpgrpc.HTTP/Handle gRPC requests into HTTP requests
+// and forwards them to its own internal HTTP router.
+//
+// By default, the server-side tracing spans for the httpgrpc.HTTP/Handle gRPC method
+// have no data about the wrapped HTTP request being handled.
+//
+// HTTPGRPCTracer.Wrap starts a child span with span name and tags following the approach in
+// Tracer.Wrap's usage of opentracing-contrib/go-stdlib/nethttp.Middleware
+// and attaches the HTTP server span tags to the parent httpgrpc.HTTP/Handle gRPC span, allowing
+// tracing tooling to differentiate the HTTP requests represented by the httpgrpc.HTTP/Handle spans.
+//
+// opentracing-contrib/go-stdlib/nethttp.Middleware could not be used here
+// as it does not expose options to access and tag the incoming parent span.
+//
+// Parent span tagging depends on using a Jaeger tracer for now to check the parent span's
+// OperationName(), which is not available on the generic opentracing Tracer interface.
+func (hgt HTTPGRPCTracer) Wrap(next http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		tracer := opentracing.GlobalTracer()
+
+		// skip spans which were not forwarded from non-httpgrpc.HTTP/Handle spans;
+		// standard http spans started directly from the HTTP server are presumed to
+		// already be instrumented by Tracer.Wrap
+		parentSpan := opentracing.SpanFromContext(ctx)
+		if parentSpan, ok := parentSpan.(*jaeger.Span); ok {
+			if parentSpan.OperationName() != httpGRPCHandleMethod {
+				next.ServeHTTP(w, r)
+				return
+			}
+		}
+
+		// extract relevant span & tag data from request
+		method := r.Method
+		matchedRoute := getRouteName(hgt.RouteMatcher, r)
+		url := r.URL.String()
+		userAgent := r.Header.Get("User-Agent")
+
+		// tag parent httpgrpc.HTTP/Handle server span, if it exists
+		if parentSpan != nil {
+			parentSpan.SetTag(string(ext.HTTPUrl), url)
+			parentSpan.SetTag(string(ext.HTTPMethod), method)
+			parentSpan.SetTag("http.route", matchedRoute)
+			parentSpan.SetTag("http.user_agent", userAgent)
+		}
+
+		// create and start child HTTP span
+		// mirroring opentracing-contrib/go-stdlib/nethttp.Middleware span name and tags
+		childSpanName := makeHTTPOperationNameFunc(hgt.RouteMatcher)(r)
+		startSpanOpts := []opentracing.StartSpanOption{
+			ext.SpanKindRPCServer,
+			opentracing.Tag{Key: string(ext.Component), Value: "net/http"},
+			opentracing.Tag{Key: string(ext.HTTPUrl), Value: url},
+			opentracing.Tag{Key: string(ext.HTTPMethod), Value: method},
+			opentracing.Tag{Key: "http.route", Value: matchedRoute},
+			opentracing.Tag{Key: "http.user_agent", Value: userAgent},
+		}
+		if parentSpan != nil {
+			startSpanOpts = append(
+				startSpanOpts,
+				opentracing.SpanReference{
+					Type:              opentracing.SpanReferenceType(1),
```

Comment:
Should we use one of the `SpanReferenceType` constants here? And should it be `ChildOfRef` (which is 0), rather than `FollowsFromRef` (1)?